src/gymFunctions.cpp

```diff
@@ -0,0 +1,169 @@
+        int argc = 0;
+        char **argv = NULL;
+        ros::init(argc, argv, "Gym functions", ros::init_options::NoSigintHandler);
+    }
+    nh = ros::NodeHandlePtr(new ros::NodeHandle);
+    spinner.reset(new ros::AsyncSpinner(0));
+    spinner->start();
+    ROS_INFO("Gym functions");
+
+    training_robot = robot;
+    training_with_limits = respect_limits;
+
+    gym_step = nh->advertiseService("/instance" + to_string(id) + "/gym_step", &gymFunctions::GymStepService,this);
+    gym_read_state = nh->advertiseService("/instance" + to_string(id) + "/gym_read_state", &gymFunctions::GymReadStateService,this);
+    gym_reset = nh->advertiseService("/instance" + to_string(id) + "/gym_reset", &gymFunctions::GymResetService,this);
+    gym_goal = nh->advertiseService("/instance" + to_string(id) + "/gym_goal", &gymFunctions::GymGoalService,this);
+    fmt = Eigen::IOFormat(4, 0, " ", ";\n", "", "", "[", "]");
+}
+
+
+bool gymFunctions::GymStepService(roboy_simulation_msgs::GymStep::Request &req,
+                                  roboy_simulation_msgs::GymStep::Response &res){
+    training_robot->update();
+
+    Map<VectorXd> action(req.set_points.data()  , training_robot->number_of_cables);
+    training_robot->Ld[0]= action;  //Commanding cable velocity for simulation
+    //training_robot->l[i] = action;     //Commanding cable length for hardware
+
+    if(!training_robot->getExternalRobotState())
+        training_robot->forwardKinematics(req.step_size);
+
+    ROS_INFO_STREAM_THROTTLE(5, "Ld = " << training_robot->Ld[0].format(fmt));
+
+    training_robot->write();
+    if(training_with_limits){
+        res.feasible = isFeasible(training_robot->getLimitVector(0), training_robot->getLimitVector(1), training_robot->q[0], training_robot->q[1]);
+        if(!res.feasible){
+            VectorXd closestLimit = findClosestJointLimit(training_robot->q); //Find closest boundary point where we can teleport
+            VectorXd jointVel;                                                //Robot hit the boundary, set velocity to zero.
+
+            jointVel.resize(training_robot->number_of_dofs);
+            jointVel.setZero();
+
+            setJointAngleAndVelocity(closestLimit, jointVel);
+        }
+    }
+    setResponse(training_robot->q,training_robot->qd,res);
+    return true;
+}
+
+bool gymFunctions::GymReadStateService(roboy_simulation_msgs::GymStep::Request &req,
+                                       roboy_simulation_msgs::GymStep::Response &res){
+
+    setResponse(training_robot->q,training_robot->qd,res);
+    if(training_with_limits)
+        res.feasible = isFeasible(training_robot->getLimitVector(0),training_robot->getLimitVector(1),training_robot->q[0],training_robot->q[1]);
+    return true;
+}
+
+bool gymFunctions::GymResetService(roboy_simulation_msgs::GymReset::Request &req,
+                                   roboy_simulation_msgs::GymReset::Response &res){
+
+    training_robot->setIntegrationTime(0.0);
```

Comment:
we set integration time to 0 and never set it back to anything else? This method seems dangerous :D

Reply:
It's set in robot.cpp. Basically step_size sets the integration_time